File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/audience_profile/compose_broadcast/ComposeBroadcastContent.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.audience_profile.compose_broadcast

import androidx.compose.runtime.Immutable
import app.pantopus.android.ui.screens.identity_center.IdentityKind
import app.pantopus.android.ui.theme.PantopusIcon

/**
 * A.7 (A22.2) Compose Broadcast — render-only models for the full-screen
 * broadcast composer pushed from the Audience Profile. Mirrors the iOS
 * `ComposeBroadcastContent.swift` shape so cross-platform parity tests can
 * compare projections one-to-one. No backend: seeded from
 * [ComposeBroadcastSampleData].
 */

/** Targeting for a broadcast — "All beacons" public reach down to tier locks. */
enum class BroadcastAudience(
    val key: String,
    val title: String,
    val icon: PantopusIcon,
    /** Persona tier rank for chip color via `tierColor(rank)`; null = public. */
    val tierRank: Int?,
) {
    AllBeacons("allBeacons", "All beacons", PantopusIcon.Globe, null),
    FollowersOnly("followersOnly", "Followers only", PantopusIcon.Users, 1),
    BronzePlus("bronzePlus", "Bronze+", PantopusIcon.Lock, 2),
    SilverPlus("silverPlus", "Silver+", PantopusIcon.Lock, 3),
    GoldOnly("goldOnly", "Gold only", PantopusIcon.Lock, 4),
    ;

    val isRestricted: Boolean get() = this != AllBeacons
}

/** One attached media item. Rendered as a tinted placeholder + caption. */
@Immutable
data class ComposeMediaPreview(
    val id: String,
    val kind: Kind,
    val caption: String?,
) {
    enum class Kind { Image, Video }
}

/** The mutable composer payload — what would be POSTed on send. */
@Immutable
data class ComposeBroadcastDraft(
    val body: String = "",
    val audience: BroadcastAudience = BroadcastAudience.AllBeacons,
    val media: ComposeMediaPreview? = null,
) {
    /** Nothing worth sending — empty body and no media. */
    val isEmpty: Boolean get() = body.isBlank() && media == null
}

/** The persona a broadcast is sent as — drives the composer PersonaRow. */
@Immutable
data class BroadcastPersona(
    val id: String,
    val handle: String,
    val displayName: String,
    val kind: IdentityKind,
    val avatarInitial: String,
)

/** One recent broadcast with inline analytics (pre-formatted strings). */
@Immutable
data class RecentBroadcastContent(
    val id: String,
    val timeLabel: String,
    val audience: BroadcastAudience,
    val body: String,
    val reach: String,
    val read: String,
    val readPct: String,
    val reactions: String,
    val replies: String,
    val hasMedia: Boolean,
)

/** Send lifecycle, kept separate from the editable draft. */
sealed interface ComposePhase {
    data object Idle : ComposePhase

    data object Sending : ComposePhase

    data class Error(val message: String) : ComposePhase
}

/**
 * The prompt's composer-state contract, derived from the live draft +
 * [ComposePhase]. Mirrors the iOS `ComposeBroadcastState` enum so parity
 * tests line up.
 */
sealed interface ComposeBroadcastState {
    data object Empty : ComposeBroadcastState

    data class Composing(val draft: ComposeBroadcastDraft) : ComposeBroadcastState

    data class Scheduled(val draft: ComposeBroadcastDraft, val sendAt: Long) : ComposeBroadcastState

    data object Sending : ComposeBroadcastState

    data class Error(val message: String) : ComposeBroadcastState
}

/**
 * Full UI state the screen renders from. The editor is always present;
 * `phase` + the draft drive [composeState] and the screen chrome.
 */
@Immutable
data class ComposeBroadcastUiState(
    val persona: BroadcastPersona,
    val recentBroadcasts: List<RecentBroadcastContent>,
    val draft: ComposeBroadcastDraft,
    val scheduledAtMillis: Long?,
    val scheduledLabel: String?,
    val phase: ComposePhase,
    val isDirty: Boolean,
    val maxCharacterCount: Int = 1_000,
    val audienceReach: Map<BroadcastAudience, Int> = emptyMap(),
) {
    val characterCount: Int get() = draft.body.length
    val isOverLimit: Boolean get() = characterCount > maxCharacterCount
    val hasRecentBroadcasts: Boolean get() = recentBroadcasts.isNotEmpty()
    val isSending: Boolean get() = phase == ComposePhase.Sending

    val canSend: Boolean
        get() = phase != ComposePhase.Sending && !draft.isEmpty && !isOverLimit

    val primaryActionTitle: String
        get() = if (hasRecentBroadcasts) "Send broadcast" else "Send your first broadcast"

    fun reach(audience: BroadcastAudience): Int? = audienceReach[audience]

    fun composeState(): ComposeBroadcastState =
        when (val current = phase) {
            ComposePhase.Sending -> ComposeBroadcastState.Sending
            is ComposePhase.Error -> ComposeBroadcastState.Error(current.message)
            ComposePhase.Idle ->
                when {
                    scheduledAtMillis != null -> ComposeBroadcastState.Scheduled(draft, scheduledAtMillis)
                    draft.isEmpty -> ComposeBroadcastState.Empty
                    else -> ComposeBroadcastState.Composing(draft)
                }
        }
}

File: frontend/apps/ios/Pantopus/Features/AudienceProfile/ComposeBroadcast/ComposeBroadcastContent.swift
//
//  ComposeBroadcastContent.swift
//  Pantopus
//
//  A.7 (A22.2) Compose Broadcast — render-only models for the
//  full-screen broadcast composer pushed from the Audience Profile.
//  The inline Audience Profile composer remains a lightweight
//  quick-post; this surface is the canonical broadcast composer
//  (large editor + media + audience targeting + scheduling + recent
//  broadcast analytics). No backend: the VM is seeded with sample
//  data matching the eventual API shapes.
//

import Foundation

/// Targeting for a broadcast. Mirrors the audience-frames TierChip map
/// ("All beacons" public reach down to per-tier locks). `tierRank` maps
/// onto the persona tier ladder so chip colors stay consistent with
/// `AudienceProfileView.tierColor(rank:)` (nil = all beacons / public).
public enum BroadcastAudience: String, Sendable, Hashable, CaseIterable, Identifiable {
    case allBeacons
    case followersOnly
    case bronzePlus
    case silverPlus
    case goldOnly

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .allBeacons: "All beacons"
        case .followersOnly: "Followers only"
        case .bronzePlus: "Bronze+"
        case .silverPlus: "Silver+"
        case .goldOnly: "Gold only"
        }
    }

    /// Lucide-token icon for the chip. Public reach gets a globe; the
    /// tier-locked options get a lock; followers gets the people glyph.
    public var icon: PantopusIcon {
        switch self {
        case .allBeacons: .globe
        case .followersOnly: .users
        case .bronzePlus, .silverPlus, .goldOnly: .lock
        }
    }

    /// Persona tier rank used to color the chip via
    /// `AudienceProfileView.tierColor(rank:)`. `nil` for all-beacons.
    public var tierRank: Int? {
        switch self {
        case .allBeacons: nil
        case .followersOnly: 1
        case .bronzePlus: 2
        case .silverPlus: 3
        case .goldOnly: 4
        }
    }

    /// True for every targeting option that isn't the public broadcast.
    public var isRestricted: Bool { self != .allBeacons }
}

/// One attached media item in the composer. `imageData` carries the
/// bytes when the user picks a real photo (rendered inline); sample /
/// snapshot data leaves it `nil` so the preview falls back to a tinted
/// placeholder — deterministic for snapshot baselines.
public struct ComposeMediaPreview: Sendable, Hashable, Identifiable {
    public enum Kind: String, Sendable, Hashable {
        case image
        case video
    }

    public let id: String
    public let kind: Kind
    public let caption: String?
    public let imageData: Data?

    public init(id: String = UUID().uuidString, kind: Kind, caption: String?, imageData: Data? = nil) {
        self.id = id
        self.kind = kind
        self.caption = caption
        self.imageData = imageData
    }
}

/// The mutable composer payload — what would be POSTed on send. Kept a
/// value type so the VM can snapshot it for the send call and compare
/// against the last-saved copy for the unsaved-draft indicator.
public struct ComposeBroadcastDraft: Sendable, Hashable {
    public var body: String
    public var audience: BroadcastAudience
    public var media: ComposeMediaPreview?

    public init(
        body: String = "",
        audience: BroadcastAudience = .allBeacons,
        media: ComposeMediaPreview? = nil
    ) {
        self.body = body
        self.audience = audience
        self.media = media
    }

    /// Nothing worth sending — empty body (whitespace-trimmed) and no
    /// media. Audience choice alone doesn't count as content.
    public var isEmpty: Bool {
        body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && media == nil
    }
}

/// The persona a broadcast is sent as. Drives the composer's PersonaRow
/// (avatar initial + handle + "Sending as <pillar>" + identity tone).
public struct BroadcastPersona: Sendable, Hashable {
    public let id: String
    public let handle: String
    public let displayName: String
    public let kind: IdentityKind
    public let avatarInitial: String

    public init(
        id: String,
        handle: String,
        displayName: String,
        kind: IdentityKind,
        avatarInitial: String
    ) {
        self.id = id
        self.handle = handle
        self.displayName = displayName
        self.kind = kind
        self.avatarInitial = avatarInitial
    }
}

/// One recent broadcast with inline analytics — borrows the audience
/// analytics inline pattern. Counts are pre-formatted display strings
/// ("1.1K", "78%") so the row stays render-only.
public struct RecentBroadcastContent: Sendable, Hashable, Identifiable {
    public let id: String
    public let timeLabel: String
    public let audience: BroadcastAudience
    public let body: String
    public let reach: String
    public let read: String
    public let readPct: String
    public let reactions: String
    public let replies: String
    public let hasMedia: Bool

    public init(
        id: String,
        timeLabel: String,
        audience: BroadcastAudience,
        body: String,
        reach: String,
        read: String,
        readPct: String,
        reactions: String,
        replies: String,
        hasMedia: Bool
    ) {
        self.id = id
        self.timeLabel = timeLabel
        self.audience = audience
        self.body = body
        self.reach = reach
        self.read = read
        self.readPct = readPct
        self.reactions = reactions
        self.replies = replies
        self.hasMedia = hasMedia
    }
}

/// Top-level composer state. The screen always renders the editor; this
/// enum tracks the composer *phase* the design's two frames + the send
/// lifecycle imply. `.empty` is the first-broadcast prompt (just opened),
/// `.composing` once there's content, `.scheduled` once a send time is
/// pinned, `.sending` mid-submit, `.error` on a failed send (the draft is
/// preserved so the user can retry).
public enum ComposeBroadcastState: Equatable, Sendable {
    case empty
    case composing(ComposeBroadcastDraft)
    case scheduled(ComposeBroadcastDraft, sendAt: Date)
    case sending
    case error(message: String)
}
